### PR TITLE
[FLINK-25249][connector/kafka] Introduce KafkaExtension for all Kafka connector tests

### PR DESCRIPTION
**Pull request by PatrickRen**
## What is the purpose of the change

This pull request introduces `KafkaExtension` as the basic test util for Kafka connector test.

This extension uses `KafkaContainer` as the underlying Kafka cluster, and provides some helper classes and functions to interact with the cluster.

## Brief change log

- Introduce KafkaExtension for all Kafka connector related tests
- Make external system optional and support static field in connector testing framework
- Use KafkaExtension in tests for KafkaSource and KafkaSink

## Verifying this change

This change is already covered by Kafka connector existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/testutils/KafkaSourceTestRecordGenerator.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import org.apache.kafka.common.serialization.IntegerSerializer;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.apache.kafka.common.serialization.StringSerializer;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
```
Comment (ruanhang1993) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/testutils/KafkaSourceTestRecordGenerator.java:17]: javadoc

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/testutils/KafkaSourceTestRecordGenerator.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+package org.apache.flink.connector.kafka.source.testutils;
```
Comment (ruanhang1993) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/testutils/KafkaSourceTestRecordGenerator.java:1]: licenses

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/testutils/cluster/KafkaContainers.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
+package org.apache.flink.connector.kafka.testutils.cluster;
```
Comment (ruanhang1993) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/testutils/cluster/KafkaContainers.java:1]: licenses

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/testutils/cluster/KafkaContainersBuilder.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
+package org.apache.flink.connector.kafka.testutils.cluster;
```
Comment (ruanhang1993) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/testutils/cluster/KafkaContainersBuilder.java:1]: licenses

**Review by JingGe — COMMENTED** [review]
Thanks for your contribution! The PR looks overall good. I just left some comments.

**Comment by fapaul** [comment]
@PatrickRen WDYT about the timeline of this PR? Currently, on the CI we see a lot of instabilities coming from Kafka legacy tests and I see these changes as a critical improvement.

**Comment by PatrickRen** [comment]
@fapaul @JingGe Sorry for my late response! I made a new push just now. Hope this round is acceptable.

Basically I changed the usage of `@Topic` annotation, and adopt advices from @JingGe. You can check the JavaDoc of new classes in the first commit to get the usage of the new infra.

Please take a look when you are available. Thanks!

**Comment by fapaul** [comment]
I had a quick glimpse and overall the structure looks fine but I was wondering if it was intended to leave all classes based on `KafkaTestBase` untouched. I would hope that using your new kit in the base class also alleviates a lot of instabilities.

**Comment by tillrohrmann** [comment]
Quick question on the state of this PR? Where are we with it and when can we probably merge it? I am noticing quite some Kafka test failures in our builds and I hope that we can solve quite a lot of them with this PR.

**Comment by PatrickRen** [comment]
@tillrohrmann Sorry for my late response! I created another simpler PR #18656 for fixing this and currently under review. I'll close this one for now.

**Comment by PatrickRen** [comment]
Please turn to #18656 for this issue

**Pull request closed without merging**